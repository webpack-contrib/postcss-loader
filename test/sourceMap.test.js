/**
 * @jest-environment node
 */

import path from "path";
import fs from "fs";

import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from "./helpers";

describe('"sourceMap" option', () => {
  it('should generate source maps with "true" value and the "devtool" with "false" value', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        sourceMap: true,
        postcssOptions: { hideNothingWarning: true },
      },
      {
        devtool: false,
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.css", stats);

    sourceMap.sourceRoot = "";
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(sourceMap.sourceRoot, source))).toBe(
        true,
      );

      return path
        .relative(path.resolve(__dirname, "./fixtures"), source)
        .replace(/\\/g, "/");
    });

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps with "true" value and the "devtool" option with "source-map" value', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        sourceMap: true,
        postcssOptions: { hideNothingWarning: true },
      },
      {
        devtool: "source-map",
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.css", stats);

    sourceMap.sourceRoot = "";
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(sourceMap.sourceRoot, source))).toBe(
        true,
      );

      return path
        .relative(path.resolve(__dirname, "./fixtures"), source)
        .replace(/\\/g, "/");
    });

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when value is not specified and the "devtool" with "source-map" value', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        postcssOptions: { hideNothingWarning: true },
      },
      {
        devtool: "source-map",
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.css", stats);

    sourceMap.sourceRoot = "";
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(sourceMap.sourceRoot, source))).toBe(
        true,
      );

      return path
        .relative(path.resolve(__dirname, "./fixtures"), source)
        .replace(/\\/g, "/");
    });

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps with "false" value, but the "postcssOptions.map" has values', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        postcssOptions: {
          map: {
            inline: false,
            annotation: false,
            prev: false,
            sourcesContent: true,
          },
          hideNothingWarning: true,
        },
      },
      {
        devtool: false,
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.css", stats);

    sourceMap.sourceRoot = "";
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(false);
      expect(source).toBe(path.normalize(source));
      expect(
        fs.existsSync(path.resolve(__dirname, "./fixtures/css", source)),
      ).toBe(true);

      return source.replace(/\\/g, "/");
    });

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps using the "postcssOptions.map" option with "true" value and previous loader returns source maps ("sass-loader")', async () => {
    const compiler = getCompiler(
      "./scss/index.js",
      {},
      {
        devtool: false,
        module: {
          rules: [
            {
              test: /\.scss$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                  options: {},
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    postcssOptions: {
                      map: true,
                      hideNothingWarning: true,
                    },
                  },
                },
                {
                  loader: "sass-loader",
                  options: {
                    // eslint-disable-next-line global-require
                    implementation: require("sass"),
                    sassOptions: {
                      sourceMap: true,
                      sourceMapRoot: path.resolve(
                        __dirname,
                        "./fixtures/scss/",
                      ),
                      outFile: path.resolve(
                        __dirname,
                        "./fixtures/scss/style.css.map",
                      ),
                      sourceMapContents: true,
                      omitSourceMapUrl: true,
                      sourceMapEmbed: false,
                    },
                  },
                },
              ],
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.scss", stats);

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps using the "postcssOptions.map" option with values and previous loader returns source maps ("sass-loader")', async () => {
    const compiler = getCompiler(
      "./scss/index.js",
      {},
      {
        devtool: false,
        module: {
          rules: [
            {
              test: /\.scss$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                  options: {},
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    postcssOptions: {
                      map: {
                        inline: false,
                        sourcesContent: true,
                        annotation: true,
                      },
                      hideNothingWarning: true,
                    },
                  },
                },
                {
                  loader: "sass-loader",
                  options: {
                    // eslint-disable-next-line global-require
                    implementation: require("sass"),
                    sassOptions: {
                      sourceMap: true,
                      sourceMapRoot: path.resolve(
                        __dirname,
                        "./fixtures/scss/",
                      ),
                      outFile: path.resolve(
                        __dirname,
                        "./fixtures/scss/style.css.map",
                      ),
                      sourceMapContents: true,
                      omitSourceMapUrl: true,
                      sourceMapEmbed: false,
                    },
                  },
                },
              ],
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.scss", stats);

    sourceMap.sourceRoot = "";
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(false);
      expect(
        fs.existsSync(path.resolve(__dirname, "./fixtures/scss", source)),
      ).toBe(true);

      return source.replace(/\\/g, "/");
    });

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps with "false" value and the "devtool" option with "false" value', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        sourceMap: false,
        postcssOptions: { hideNothingWarning: true },
      },
      {
        devtool: false,
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.css", stats);

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps with "false" value and the "devtool" option with "source-map" value', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        sourceMap: false,
        postcssOptions: { hideNothingWarning: true },
      },
      {
        devtool: "source-map",
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.css", stats);

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps when value is not specified and the "devtool" option with "source-map" value', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        postcssOptions: { hideNothingWarning: true },
      },
      {
        devtool: false,
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.css", stats);

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when previous loader returns source maps ("sass-loader")', async () => {
    const compiler = getCompiler(
      "./scss/index.js",
      {},
      {
        devtool: "source-map",
        module: {
          rules: [
            {
              test: /\.scss$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    postcssOptions: { hideNothingWarning: true },
                  },
                },
                {
                  loader: "sass-loader",
                  options: {
                    // eslint-disable-next-line global-require
                    implementation: require("sass"),
                  },
                },
              ],
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.scss", stats);

    sourceMap.sourceRoot = "";
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(sourceMap.sourceRoot, source))).toBe(
        true,
      );

      return path
        .relative(path.resolve(__dirname, "./fixtures"), source)
        .replace(/\\/g, "/");
    });

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when previous loader returns source maps ("less-loader")', async () => {
    const compiler = getCompiler(
      "./less/index.js",
      {},
      {
        devtool: "source-map",
        module: {
          rules: [
            {
              test: /\.less$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                  options: {},
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    postcssOptions: { hideNothingWarning: true },
                  },
                },
                {
                  loader: "less-loader",
                },
              ],
            },
          ],
        },
      },
    );
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle("style.less", stats);

    sourceMap.sourceRoot = "";
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(sourceMap.sourceRoot, source))).toBe(
        true,
      );

      return path
        .relative(path.resolve(__dirname, "./fixtures"), source)
        .replace(/\\/g, "/");
    });

    expect(css).toMatchSnapshot("css");
    expect(sourceMap).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate inline source maps when the "devtool" is "false"', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        postcssOptions: {
          map: {
            inline: true,
            annotation: false,
          },
          hideNothingWarning: true,
        },
      },
      {
        devtool: false,
      },
    );
    const stats = await compile(compiler);
    const { css } = getCodeFromBundle("style.css", stats);

    expect(css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate inline source maps when the "devtool" is "true"', async () => {
    const compiler = getCompiler(
      "./css/index.js",
      {
        postcssOptions: {
          map: {
            inline: true,
            annotation: false,
          },
          hideNothingWarning: true,
        },
      },
      {
        devtool: "source-map",
      },
    );
    const stats = await compile(compiler);
    const { css } = getCodeFromBundle("style.css", stats);

    expect(css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
