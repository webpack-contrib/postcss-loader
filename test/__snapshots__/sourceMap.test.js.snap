// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"sourceMap" option should generate inline source maps when the "devtool" is "false": css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInN0eWxlLmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSIsImZpbGUiOiJzdHlsZS5jc3MiLCJzb3VyY2VzQ29udGVudCI6WyJhIHtcbiAgY29sb3I6IGJsYWNrO1xufVxuXG5hIHtcbiAgY29sb3I6IHJlZDtcbn1cblxuYSB7XG4gIGNvbG9yOiBncmVlbjtcbn1cblxuYSB7XG4gIGNvbG9yOiBibHVlO1xufVxuXG4uY2xhc3Mge1xuICAteC1ib3JkZXItY29sb3I6IGJsdWUgYmx1ZSAqO1xuICAteC1jb2xvcjogKiAjZmFmYWZhO1xufVxuXG4uY2xhc3MtZm9vIHtcbiAgLXotYm9yZGVyLWNvbG9yOiBibHVlIGJsdWUgKjtcbiAgLXotY29sb3I6ICogI2ZhZmFmYTtcbn1cblxuLnBob25lIHtcbiAgJl90aXRsZSB7XG4gICAgd2lkdGg6IDUwMHB4O1xuXG4gICAgQG1lZGlhIChtYXgtd2lkdGg6IDUwMHB4KSB7XG4gICAgICB3aWR0aDogYXV0bztcbiAgICB9XG5cbiAgICBib2R5LmlzX2RhcmsgJiB7XG4gICAgICBjb2xvcjogd2hpdGU7XG4gICAgfVxuICB9XG5cbiAgaW1nIHtcbiAgICBkaXNwbGF5OiBibG9jaztcbiAgfVxufVxuIl19 */"
`;

exports[`"sourceMap" option should generate inline source maps when the "devtool" is "false": errors 1`] = `[]`;

exports[`"sourceMap" option should generate inline source maps when the "devtool" is "false": warnings 1`] = `[]`;

exports[`"sourceMap" option should generate inline source maps when the "devtool" is "true": css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInN0eWxlLmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSIsImZpbGUiOiJzdHlsZS5jc3MiLCJzb3VyY2VzQ29udGVudCI6WyJhIHtcbiAgY29sb3I6IGJsYWNrO1xufVxuXG5hIHtcbiAgY29sb3I6IHJlZDtcbn1cblxuYSB7XG4gIGNvbG9yOiBncmVlbjtcbn1cblxuYSB7XG4gIGNvbG9yOiBibHVlO1xufVxuXG4uY2xhc3Mge1xuICAteC1ib3JkZXItY29sb3I6IGJsdWUgYmx1ZSAqO1xuICAteC1jb2xvcjogKiAjZmFmYWZhO1xufVxuXG4uY2xhc3MtZm9vIHtcbiAgLXotYm9yZGVyLWNvbG9yOiBibHVlIGJsdWUgKjtcbiAgLXotY29sb3I6ICogI2ZhZmFmYTtcbn1cblxuLnBob25lIHtcbiAgJl90aXRsZSB7XG4gICAgd2lkdGg6IDUwMHB4O1xuXG4gICAgQG1lZGlhIChtYXgtd2lkdGg6IDUwMHB4KSB7XG4gICAgICB3aWR0aDogYXV0bztcbiAgICB9XG5cbiAgICBib2R5LmlzX2RhcmsgJiB7XG4gICAgICBjb2xvcjogd2hpdGU7XG4gICAgfVxuICB9XG5cbiAgaW1nIHtcbiAgICBkaXNwbGF5OiBibG9jaztcbiAgfVxufVxuIl19 */"
`;

exports[`"sourceMap" option should generate inline source maps when the "devtool" is "true": errors 1`] = `[]`;

exports[`"sourceMap" option should generate inline source maps when the "devtool" is "true": warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps using the "postcssOptions.map" option with "true" value and previous loader returns source maps ("sass-loader"): css 1`] = `
"a {
  color: coral;
}
/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInN0eWxlLnNjc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7RUFBSSIsImZpbGUiOiJzdHlsZS5zY3NzIiwic291cmNlc0NvbnRlbnQiOlsiYSB7XG4gIGNvbG9yOiBjb3JhbDtcbn0iXX0= */"
`;

exports[`"sourceMap" option should generate source maps using the "postcssOptions.map" option with "true" value and previous loader returns source maps ("sass-loader"): errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps using the "postcssOptions.map" option with "true" value and previous loader returns source maps ("sass-loader"): warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps using the "postcssOptions.map" option with values and previous loader returns source maps ("sass-loader"): css 1`] = `
"a {
  color: coral;
}
/*# sourceMappingURL=style.scss.map */"
`;

exports[`"sourceMap" option should generate source maps using the "postcssOptions.map" option with values and previous loader returns source maps ("sass-loader"): errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps using the "postcssOptions.map" option with values and previous loader returns source maps ("sass-loader"): source map 1`] = `
{
  "file": "style.scss",
  "mappings": "AAAA;EAAI",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "style.scss",
  ],
  "sourcesContent": [
    "a {
  color: coral;
}",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" option should generate source maps using the "postcssOptions.map" option with values and previous loader returns source maps ("sass-loader"): warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("less-loader"): css 1`] = `
"a {
  color: coral;
}
"
`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("less-loader"): errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("less-loader"): source map 1`] = `
{
  "mappings": "AAAA;EAAI,YAAA",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "less/style.less",
  ],
  "sourcesContent": [
    "a {
  color: coral;
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("less-loader"): warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("sass-loader"): css 1`] = `
"a {
  color: coral;
}"
`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("sass-loader"): errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("sass-loader"): source map 1`] = `
{
  "mappings": "AAAA;EAAI",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "scss/style.scss",
  ],
  "sourcesContent": [
    "a {
  color: coral;
}",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" option should generate source maps when previous loader returns source maps ("sass-loader"): warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps when value is not specified and the "devtool" with "source-map" value: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"sourceMap" option should generate source maps when value is not specified and the "devtool" with "source-map" value: errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps when value is not specified and the "devtool" with "source-map" value: source map 1`] = `
{
  "mappings": "AAAA",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "css/style.css",
  ],
  "sourcesContent": [
    "a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" option should generate source maps when value is not specified and the "devtool" with "source-map" value: warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps with "false" value, but the "postcssOptions.map" has values: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"sourceMap" option should generate source maps with "false" value, but the "postcssOptions.map" has values: errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps with "false" value, but the "postcssOptions.map" has values: source map 1`] = `
{
  "file": "style.css",
  "mappings": "AAAA",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "style.css",
  ],
  "sourcesContent": [
    "a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" option should generate source maps with "false" value, but the "postcssOptions.map" has values: warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" option with "source-map" value: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" option with "source-map" value: errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" option with "source-map" value: source map 1`] = `
{
  "mappings": "AAAA",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "css/style.css",
  ],
  "sourcesContent": [
    "a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" option with "source-map" value: warnings 1`] = `[]`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" with "false" value: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" with "false" value: errors 1`] = `[]`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" with "false" value: source map 1`] = `
{
  "mappings": "AAAA",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "css/style.css",
  ],
  "sourcesContent": [
    "a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" option should generate source maps with "true" value and the "devtool" with "false" value: warnings 1`] = `[]`;

exports[`"sourceMap" option should not generate source maps when value is not specified and the "devtool" option with "source-map" value: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"sourceMap" option should not generate source maps when value is not specified and the "devtool" option with "source-map" value: errors 1`] = `[]`;

exports[`"sourceMap" option should not generate source maps when value is not specified and the "devtool" option with "source-map" value: warnings 1`] = `[]`;

exports[`"sourceMap" option should not generate source maps with "false" value and the "devtool" option with "false" value: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"sourceMap" option should not generate source maps with "false" value and the "devtool" option with "false" value: errors 1`] = `[]`;

exports[`"sourceMap" option should not generate source maps with "false" value and the "devtool" option with "false" value: warnings 1`] = `[]`;

exports[`"sourceMap" option should not generate source maps with "false" value and the "devtool" option with "source-map" value: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"sourceMap" option should not generate source maps with "false" value and the "devtool" option with "source-map" value: errors 1`] = `[]`;

exports[`"sourceMap" option should not generate source maps with "false" value and the "devtool" option with "source-map" value: warnings 1`] = `[]`;
